File: main/reconstruct-itinerary/reconstruct-itinerary-fast.cs
public class Solution {
    public IList<string> FindItinerary(string[,] tickets) {
        // Create the adjacency list.
        var adj = new Dictionary<string, List<string>>();
        foreach (var i in Enumerable.Range(0, tickets.GetLength(0))) {
            var src = tickets[i, 0];
            if (!adj.TryGetValue(src, out var row)) {
                row = new List<String>();
                adj.Add(src, row);
            }
            row.Add(tickets[i, 1]);
        }

        // Sort each row so items can be popped in lexicographic order.
        foreach (var row in adj.Values) row.Sort((s, t) => t.CompareTo(s));

        // We will build the path backwards.
        var path = new List<string>();

        // Recursive Hierholzer's algorithm.
        void Dfs(string src) {
            if (adj.TryGetValue(src, out var row)) {
                while (row.Count != 0) {
                    var dest = row[row.Count - 1];
                    row.RemoveAt(row.Count - 1);
                    Dfs(dest);
                }
            }

            path.Add(src);
        }

        // Build the path and return it forwards.
        Dfs("JFK");
        path.Reverse();
        return path;
    }
}

File: main/reconstruct-itinerary/reconstruct-itinerary-fast.py
class Solution:
    def findItinerary(self, tickets):
        """
        :type tickets: List[List[str]]
        :rtype: List[str]
        """
        # Create the adjacency list.
        adj = collections.defaultdict(list)
        for src, dest in tickets:
            adj[src].append(dest)

        # Sort each row so items can be popped in lexicographic order.
        for row in adj.values():
            row.sort(reverse=True)

        # We will build the path backwards.
        path = []

        # Recursive Hierholzer's algorithm.
        def dfs(src):
            row = adj[src]
            while row:
                dfs(row.pop())
            path.append(src)

        # Build the path and return it forwards.
        dfs('JFK')
        path.reverse()
        return path
